lib/cylc/ws_messages.proto

```diff
@@ -0,0 +1,227 @@
+    repeated string states = 27;
+}
+
+message PbJob {
+    string checksum = 1;
+    string id = 2;
+    int32 submit_num = 3;
+    string state = 4;
+    string task_proxy = 5;
+    string submitted_time = 6;
+    string started_time = 7;
+    string finished_time = 8;
+    string batch_sys_job_id = 9;
+    string batch_sys_name = 10;
+    string env_script = 11;
+    string err_script = 12;
+    string exit_script = 13;
+    float execution_time_limit = 14;
+    string host = 15;
+    string init_script = 16;
+    string job_log_dir = 17;
+    string owner = 18;
+    string post_script = 19;
+    string pre_script = 20;
+    string script = 21;
+    string shell = 22;
+    string work_sub_dir = 23;
+    repeated string batch_sys_conf = 24;
+    repeated string environment = 25;
+    repeated string directives = 26;
+    repeated string param_env_tmpl = 27;
+    repeated string param_var = 28;
+    repeated string extra_logs = 29;
+}
+
+message PbTask {
+    string checksum = 1;
+    string id = 2;
+    string name = 3;
+    PbMeta meta = 4;
+    float mean_elapsed_time = 5;
+    int32 depth = 6;
+    repeated string proxies = 7;
+    repeated string namespace = 8;
+}
+
+message PbPollTask {
+    string local_proxy = 1;
+    string workflow = 2;
+    string remote_proxy = 3;
+    string req_state = 4;
+    string graph_string = 5;
+}
+
+message PbCondition {
+    string task_id = 1;
+    string task_proxy = 2;
+    string expr_alias = 3;
+    string req_state = 4;
+    bool satisfied = 5;
+    string message = 6;
+}
+
+message PbPrerequisite {
+    string expression = 1;
+    repeated PbCondition conditions = 2;
+    repeated string cycle_points = 3;
+    bool satisfied = 4;
+}
+
+message PbTaskProxy {
+    string checksum = 1;
+    string id = 2;
+    string task = 3;
+    string state = 4;
+    string cycle_point = 5;
+    bool spawned = 6;
+    int32 depth = 7;
+    int32 job_submits = 8;
+    string latest_message = 9;
+    repeated string outputs = 10;
+    repeated string broadcasts = 11;
+    repeated string namespace = 12;
+    repeated string proxy_namespace = 13;
+    repeated PbPrerequisite prerequisites = 14;
+    repeated string jobs = 15;
+    repeated string parents = 16;
+}
+
+message PbFamily {
+    string checksum = 1;
+    string id = 2;
+    string name = 3;
+    PbMeta meta = 4;
+    int32 depth = 5;
+    repeated string proxies = 6;
+    repeated string parents = 7;
+    repeated string child_tasks = 8;
+    repeated string child_families = 9;
+}
+
+message PbFamilyProxy {
+    string checksum = 1;
+    string id = 2;
+    string cycle_point = 3;
+    string name = 4;
+    string family = 5;
+    string state = 6;
+    int32 depth = 7;
+    repeated string parents = 8;
+    repeated string child_tasks = 9;
+    repeated string child_families = 10;
+}
+
+message PbEdge {
+    string id = 2;
+    string tail_node = 3;
+    string head_node = 4;
+    bool suicide = 5;
+    bool cond = 6;
+}
+
+message PbEdges {
```

Comment:
Just for curiosity, we have `Edge` and `Edges. Then a `EntireWorkflow` has a single `Edge` called `edges`. While a `Workflow` has `edges` field which is a `Pages` type. Q/ Why `EntireWorkflow`'s `edges` is a `Edge` and not a `Edges`?

Reply:
Because the edges contained in `Edges` is just an ID (string) reference, and not the real thing. and `EntireWorkflow` has a list of the objects.. Relationships ID so less repeated data.